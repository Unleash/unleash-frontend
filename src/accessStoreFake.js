import { Map as $MAp } from 'immutable';

export const createFakeStore = (permissions) => {
    return {
        getState: () => ({
            user: 
                new $MAp({
                    permissions
                })
        }),
    }
}