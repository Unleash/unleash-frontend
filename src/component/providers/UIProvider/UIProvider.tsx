import React, { useState } from 'react';
import UIContext, { createEmptyToast } from '../../../contexts/UIContext';
import { IToast } from '../../../interfaces/toast';

const UIProvider: React.FC = ({ children }) => {
    const [toastData, setToast] = useState<IToast>(createEmptyToast());
    const [showFeedback, setShowFeedback] = useState(false);

    const context = React.useMemo(
        () => ({
            setToast,
            toastData,
            showFeedback,
            setShowFeedback,
        }),
        [toastData, showFeedback]
    );

    return <UIContext.Provider value={context}>{children}</UIContext.Provider>;
};

export default UIProvider;
