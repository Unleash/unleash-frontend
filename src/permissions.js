export const ADMIN = 'ADMIN';
export const CREATE_FEATURE = 'CREATE_FEATURE';
export const UPDATE_FEATURE = 'UPDATE_FEATURE';
export const DELETE_FEATURE = 'DELETE_FEATURE';
export const CREATE_STRATEGY = 'CREATE_STRATEGY';
export const UPDATE_STRATEGY = 'UPDATE_STRATEGY';
export const DELETE_STRATEGY = 'DELETE_STRATEGY';
export const UPDATE_APPLICATION = 'UPDATE_APPLICATION';
export const CREATE_CONTEXT_FIELD = 'CREATE_CONTEXT_FIELD';
export const UPDATE_CONTEXT_FIELD = 'UPDATE_CONTEXT_FIELD';
export const DELETE_CONTEXT_FIELD = 'DELETE_CONTEXT_FIELD';

export function hasPermission(user, permission) {
    return (
        user &&
        (!user.permissions || user.permissions.indexOf(ADMIN) !== -1 || user.permissions.indexOf(permission) !== -1)
    );
}
