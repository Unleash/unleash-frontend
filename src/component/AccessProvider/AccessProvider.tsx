import { FC } from "react";

import AccessContext from '../../contexts/AccessContext'
import { ADMIN } from "./permissions";

// TODO: Type up redux store
interface IAccessProvider {
    store: any;
}

interface IPermission {
  permission: string;
  project: string | null;
}

const AccessProvider: FC<IAccessProvider> = ({store, children}) => {
  const hasAccess = (permission: string, project: string) => {
    const permissions = store.getState().user.get('permissions') || [];

    const result = permissions.some((p: IPermission) => {
      if(p.permission === ADMIN) {
        return true
      }
      if(p.permission === permission && p.project === project) {
        return true;
      }
      return false;
    });

    return result;
  };

  const context = { hasAccess };

  return <AccessContext.Provider value={context}>{children}</AccessContext.Provider>
}

export default AccessProvider;