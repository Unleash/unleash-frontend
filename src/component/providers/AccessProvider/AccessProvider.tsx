import { ReactElement, ReactNode, useCallback, useMemo } from 'react';
import AccessContext, { IAccessContext } from '../../../contexts/AccessContext';
import { ADMIN } from './permissions';
import { IPermission } from '../../../interfaces/user';
import { useAuthPermissions } from '../../../hooks/api/getters/useAuth/useAuthPermissions';

interface IAccessProviderProps {
    children: ReactNode;
    permissions: IPermission[];
}

// TODO(olav): Mock useAuth instead of using props.permissions in tests.
const AccessProvider = (props: IAccessProviderProps): ReactElement => {
    const auth = useAuthPermissions();
    const permissions = props.permissions ?? auth.permissions;

    const isAdmin: boolean = useMemo(() => {
        return checkAdmin(permissions);
    }, [permissions]);

    const hasAccess = useCallback(
        (permission: string, project?: string, environment?: string) => {
            return checkPermissions(
                permissions,
                permission,
                project,
                environment
            );
        },
        [permissions]
    );

    const value: IAccessContext = useMemo(
        () => ({
            isAdmin,
            hasAccess,
        }),
        [isAdmin, hasAccess]
    );

    return (
        <AccessContext.Provider value={value}>
            {props.children}
        </AccessContext.Provider>
    );
};

const checkAdmin = (permissions: IPermission[] | undefined): boolean => {
    if (!permissions) {
        return false;
    }

    return permissions.some(p => {
        return p.permission === ADMIN;
    });
};

const checkPermissions = (
    permissions: IPermission[] | undefined,
    permission: string,
    project?: string,
    environment?: string
): boolean => {
    if (!permissions) {
        return false;
    }

    return permissions.some(p => {
        return checkPermission(p, permission, project, environment);
    });
};

const checkPermission = (
    p: IPermission,
    permission: string,
    project?: string,
    environment?: string
): boolean => {
    if (!permission) {
        console.warn(`Missing permission for AccessProvider: ${permission}`)
        return false
    }

    if (p.permission === ADMIN) {
        return true;
    }

    if (
        p.permission === permission &&
        (p.project === project || p.project === '*') &&
        (p.environment === environment || p.environment === '*')
    ) {
        return true;
    }

    if (
        p.permission === permission &&
        (p.project === project || p.project === '*') &&
        p.environment === null
    ) {
        return true;
    }

    return (
        p.permission === permission &&
        p.project === undefined &&
        p.environment === null
    );
};

export default AccessProvider;
