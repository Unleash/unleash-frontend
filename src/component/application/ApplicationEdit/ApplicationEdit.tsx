/* eslint react/no-multi-comp:off */
import React, { useContext, useState } from 'react';
import {
    Avatar,
    Icon,
    IconButton,
    LinearProgress,
    Link,
    Typography,
} from '@material-ui/core';
import { Link as LinkIcon } from '@material-ui/icons';
import ConditionallyRender from 'component/common/ConditionallyRender/ConditionallyRender';
import { UPDATE_APPLICATION } from 'component/providers/AccessProvider/permissions';
import { ApplicationView } from '../ApplicationView/ApplicationView';
import { ApplicationUpdate } from '../ApplicationUpdate/ApplicationUpdate';
import Dialogue from 'component/common/Dialogue';
import PageContent from 'component/common/PageContent';
import HeaderTitle from 'component/common/HeaderTitle';
import AccessContext from 'contexts/AccessContext';
import useApplicationsApi from 'hooks/api/actions/useApplicationsApi/useApplicationsApi';
import useApplication from 'hooks/api/getters/useApplication/useApplication';
import { useHistory, useParams } from 'react-router-dom';
import { useLocationSettings } from 'hooks/useLocationSettings';
import useToast from 'hooks/useToast';
import PermissionButton from 'component/common/PermissionButton/PermissionButton';
import { formatDateYMD } from 'utils/formatDate';
import { formatUnknownError } from 'utils/formatUnknownError';
import { TabNav } from 'component/common/TabNav/TabNav/TabNav';

export const ApplicationEdit = () => {
    const history = useHistory();
    const { name } = useParams<{ name: string }>();
    const { application, loading } = useApplication(name);
    const { appName, url, description, icon = 'apps', createdAt } = application;
    const { hasAccess } = useContext(AccessContext);
    const { deleteApplication } = useApplicationsApi();
    const { locationSettings } = useLocationSettings();
    const { setToastData, setToastApiError } = useToast();

    const [showDialog, setShowDialog] = useState(false);

    const toggleModal = () => {
        setShowDialog(!showDialog);
    };

    const formatDate = (v: string) => formatDateYMD(v, locationSettings.locale);

    const onDeleteApplication = async (evt: React.SyntheticEvent) => {
        evt.preventDefault();
        try {
            await deleteApplication(appName);
            setToastData({
                title: 'Deleted Successfully',
                text: 'Application deleted successfully',
                type: 'success',
            });
            history.push('/applications');
        } catch (error: unknown) {
            setToastApiError(formatUnknownError(error));
        }
    };

    const renderModal = () => (
        <Dialogue
            open={showDialog}
            onClose={toggleModal}
            onClick={onDeleteApplication}
            title="Are you sure you want to delete this application?"
        />
    );
    const tabData = [
        {
            label: 'Application overview',
            component: <ApplicationView />,
        },
        {
            label: 'Edit application',
            component: <ApplicationUpdate application={application} />,
        },
    ];

    if (loading) {
        return (
            <div>
                <p>Loading...</p>
                <LinearProgress />
            </div>
        );
    } else if (!application) {
        return <p>Application ({appName}) not found</p>;
    }
    return (
        <PageContent
            headerContent={
                <HeaderTitle
                    title={
                        <span
                            style={{
                                display: 'flex',
                                alignItems: 'center',
                            }}
                        >
                            <Avatar style={{ marginRight: '8px' }}>
                                <Icon>{icon || 'apps'}</Icon>
                            </Avatar>
                            {appName}
                        </span>
                    }
                    actions={
                        <>
                            <ConditionallyRender
                                condition={Boolean(url)}
                                show={
                                    <IconButton component={Link} href={url}>
                                        <LinkIcon titleAccess={url} />
                                    </IconButton>
                                }
                            />

                            <PermissionButton
                                tooltip="Delete application"
                                onClick={toggleModal}
                                permission={UPDATE_APPLICATION}
                            >
                                Delete
                            </PermissionButton>
                        </>
                    }
                />
            }
        >
            <div>
                <Typography variant="body1">{description || ''}</Typography>
                <Typography variant="body2">
                    Created: <strong>{formatDate(createdAt)}</strong>
                </Typography>
            </div>
            <ConditionallyRender
                condition={hasAccess(UPDATE_APPLICATION)}
                show={
                    <div>
                        {renderModal()}
                        <TabNav tabData={tabData} />
                    </div>
                }
            />
        </PageContent>
    );
};
