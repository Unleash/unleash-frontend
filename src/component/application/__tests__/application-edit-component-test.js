import { ThemeProvider } from '@material-ui/core';
import { ApplicationEdit } from '../ApplicationEdit/ApplicationEdit';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { ADMIN } from '../../providers/AccessProvider/permissions';
import theme from '../../../themes/main-theme';
import { createFakeStore } from '../../../accessStoreFake';
import AccessProvider from '../../providers/AccessProvider/AccessProvider';
import UIProvider from '../../providers/UIProvider/UIProvider';

test('renders correctly if no application', () => {
    const tree = renderer
        .create(
            <AccessProvider permissions={[{ permission: ADMIN }]}>
                <ThemeProvider theme={theme}>
                    <UIProvider>
                        <MemoryRouter initialEntries={['/test']}>
                            <ApplicationEdit
                                fetchApplication={() => Promise.resolve({})}
                                storeApplicationMetaData={jest.fn()}
                                deleteApplication={jest.fn()}
                                history={{}}
                                locationSettings={{ locale: 'en-GB' }}
                            />
                        </MemoryRouter>
                    </UIProvider>
                </ThemeProvider>
            </AccessProvider>
        )
        .toJSON();

    expect(tree).toMatchSnapshot();
});

test('renders correctly without permission', () => {
    const tree = renderer
        .create(
            <MemoryRouter>
                <ThemeProvider theme={theme}>
                    <UIProvider>
                        <AccessProvider permissions={[{ permission: ADMIN }]}>
                            <ApplicationEdit
                                fetchApplication={() => Promise.resolve({})}
                                storeApplicationMetaData={jest.fn()}
                                deleteApplication={jest.fn()}
                                history={{}}
                                application={{
                                    appName: 'test-app',
                                    instances: [
                                        {
                                            instanceId: 'instance-1',
                                            clientIp: '123.123.123.123',
                                            lastSeen: '2017-02-23T15:56:49',
                                            sdkVersion: '4.0',
                                        },
                                    ],
                                    strategies: [
                                        {
                                            name: 'StrategyA',
                                            description: 'A description',
                                        },
                                        {
                                            name: 'StrategyB',
                                            description: 'B description',
                                            notFound: true,
                                        },
                                    ],
                                    seenToggles: [
                                        {
                                            name: 'ToggleA',
                                            description: 'this is A toggle',
                                            enabled: true,
                                            project: 'default',
                                        },
                                        {
                                            name: 'ToggleB',
                                            description: 'this is B toggle',
                                            enabled: false,
                                            notFound: true,
                                            project: 'default',
                                        },
                                    ],
                                    url: 'http://example.org',
                                    description: 'app description',
                                }}
                                locationSettings={{ locale: 'en-GB' }}
                            />
                        </AccessProvider>
                    </UIProvider>
                </ThemeProvider>
            </MemoryRouter>
        )
        .toJSON();

    expect(tree).toMatchSnapshot();
});

test('renders correctly with permissions', () => {
    const tree = renderer
        .create(
            <MemoryRouter>
                <ThemeProvider theme={theme}>
                    <UIProvider>
                        <AccessProvider permissions={[{ permission: ADMIN }]}>
                            <ApplicationEdit
                                fetchApplication={() => Promise.resolve({})}
                                storeApplicationMetaData={jest.fn()}
                                history={{}}
                                deleteApplication={jest.fn()}
                                application={{
                                    appName: 'test-app',
                                    instances: [
                                        {
                                            instanceId: 'instance-1',
                                            clientIp: '123.123.123.123',
                                            lastSeen: '2017-02-23T15:56:49',
                                            sdkVersion: '4.0',
                                        },
                                    ],
                                    strategies: [
                                        {
                                            name: 'StrategyA',
                                            description: 'A description',
                                        },
                                        {
                                            name: 'StrategyB',
                                            description: 'B description',
                                            notFound: true,
                                        },
                                    ],
                                    seenToggles: [
                                        {
                                            name: 'ToggleA',
                                            description: 'this is A toggle',
                                            enabled: true,
                                            project: 'default',
                                        },
                                        {
                                            name: 'ToggleB',
                                            description: 'this is B toggle',
                                            enabled: false,
                                            notFound: true,
                                            project: 'default',
                                        },
                                    ],
                                    url: 'http://example.org',
                                    description: 'app description',
                                }}
                                locationSettings={{ locale: 'en-GB' }}
                            />
                        </AccessProvider>
                    </UIProvider>
                </ThemeProvider>
            </MemoryRouter>
        )
        .toJSON();

    expect(tree).toMatchSnapshot();
});
