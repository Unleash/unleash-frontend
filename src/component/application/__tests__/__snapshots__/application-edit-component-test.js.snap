// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly if no application 1`] = `
<div>
  <p>
    Loading...
  </p>
  <div
    className="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate"
    role="progressbar"
  >
    <div
      className="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate"
      style={Object {}}
    />
    <div
      className="MuiLinearProgress-bar MuiLinearProgress-bar2Indeterminate MuiLinearProgress-barColorPrimary"
      style={Object {}}
    />
  </div>
</div>
`;

exports[`renders correctly with permissions 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  style={
    Object {
      "borderRadius": "10px",
      "boxShadow": "none",
    }
  }
>
  <div
    className="makeStyles-headerContainer-1"
  >
    <div
      className="makeStyles-headerTitleContainer-5"
    >
      <div
        className=""
        data-loading={true}
      >
        <h2
          className="MuiTypography-root makeStyles-headerTitle-6 MuiTypography-h2"
        >
          <span
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
              }
            }
          >
            <div
              className="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault"
              style={
                Object {
                  "marginRight": "8px",
                }
              }
            >
              <span
                aria-hidden={true}
                className="material-icons MuiIcon-root"
              >
                apps
              </span>
            </div>
            test-app
          </span>
        </h2>
      </div>
      <div
        className="makeStyles-headerActions-7"
      >
        <a
          aria-disabled={false}
          className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiButtonBase-root MuiIconButton-root MuiTypography-colorPrimary"
          href="http://example.org"
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z"
              />
            </svg>
          </span>
        </a>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSecondary"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          title="Delete application"
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Delete
          </span>
        </button>
      </div>
    </div>
  </div>
  <div
    className="makeStyles-bodyContainer-2"
  >
    <div>
      <p
        className="MuiTypography-root MuiTypography-body1"
      >
        app description
      </p>
      <p
        className="MuiTypography-root MuiTypography-body2"
      >
        Created: 
        <strong>
          Invalid Date
        </strong>
      </p>
    </div>
    <div>
      <div
        className="MuiPaper-root makeStyles-tabNav-10 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          className="MuiTabs-root"
        >
          <div
            className="MuiTabs-scroller MuiTabs-fixed"
            onScroll={[Function]}
            style={
              Object {
                "marginBottom": null,
                "overflow": "hidden",
              }
            }
          >
            <div
              className="MuiTabs-flexContainer MuiTabs-centered"
              onKeyDown={[Function]}
              role="tablist"
            >
              <button
                aria-controls="tabpanel-0"
                aria-selected={true}
                className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected"
                disabled={false}
                id="tab-0"
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                role="tab"
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiTab-wrapper"
                >
                  Application overview
                </span>
                <span
                  className="PrivateTabIndicator-root-11 PrivateTabIndicator-colorPrimary-12 MuiTabs-indicator"
                  style={Object {}}
                />
              </button>
              <button
                aria-controls="tabpanel-1"
                aria-selected={false}
                className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary"
                disabled={false}
                id="tab-1"
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                role="tab"
                tabIndex={-1}
                type="button"
              >
                <span
                  className="MuiTab-wrapper"
                >
                  Edit application
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          aria-labelledby="wrapped-tab-0"
          hidden={false}
          id="wrapped-tabpanel-0"
          role="tabpanel"
        >
          <div
            className="MuiGrid-root MuiGrid-container"
            style={
              Object {
                "margin": 0,
              }
            }
          >
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 MuiGrid-grid-xl-6"
            >
              <h6
                className="MuiTypography-root MuiTypography-subtitle1"
                style={
                  Object {
                    "padding": "1rem 0",
                  }
                }
              >
                Toggles
              </h6>
              <hr />
              <ul
                className="MuiList-root MuiList-padding"
              >
                <li
                  className="MuiListItem-root MuiListItem-gutters"
                  disabled={false}
                >
                  <div
                    className="MuiListItemAvatar-root"
                  >
                    <span
                      className="MuiSwitch-root"
                    >
                      <span
                        aria-disabled={true}
                        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-15 MuiSwitch-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-disabled-17 Mui-disabled Mui-disabled Mui-disabled"
                        onBlur={[Function]}
                        onDragLeave={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                        onMouseDown={[Function]}
                        onMouseLeave={[Function]}
                        onMouseUp={[Function]}
                        onTouchEnd={[Function]}
                        onTouchMove={[Function]}
                        onTouchStart={[Function]}
                        tabIndex={-1}
                      >
                        <span
                          className="MuiIconButton-label"
                        >
                          <input
                            className="PrivateSwitchBase-input-18 MuiSwitch-input"
                            disabled={true}
                            onChange={[Function]}
                            type="checkbox"
                            value={true}
                          />
                          <span
                            className="MuiSwitch-thumb"
                          />
                        </span>
                      </span>
                      <span
                        className="MuiSwitch-track"
                      />
                    </span>
                  </div>
                  <div
                    className="MuiListItemText-root MuiListItemText-multiline"
                  >
                    <span
                      className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                    >
                      <a
                        href="/projects/default/features/ToggleA/strategies/ToggleA"
                        onClick={[Function]}
                      >
                        ToggleA
                      </a>
                    </span>
                    <p
                      className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                    >
                      this is A toggle
                    </p>
                  </div>
                </li>
                <li
                  className="MuiListItem-root MuiListItem-gutters"
                  disabled={false}
                >
                  <div
                    className="MuiListItemAvatar-root"
                  >
                    <svg
                      aria-hidden={true}
                      className="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M15.73 3H8.27L3 8.27v7.46L8.27 21h7.46L21 15.73V8.27L15.73 3zM12 17.3c-.72 0-1.3-.58-1.3-1.3 0-.72.58-1.3 1.3-1.3.72 0 1.3.58 1.3 1.3 0 .72-.58 1.3-1.3 1.3zm1-4.3h-2V7h2v6z"
                      />
                    </svg>
                  </div>
                  <div
                    className="MuiListItemText-root MuiListItemText-multiline"
                  >
                    <span
                      className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                    >
                      <a
                        href="/projects/default/create-toggle?name=ToggleB"
                        onClick={[Function]}
                      >
                        ToggleB
                      </a>
                    </span>
                    <p
                      className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                    >
                      Missing, want to create?
                    </p>
                  </div>
                </li>
              </ul>
            </div>
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 MuiGrid-grid-xl-6"
            >
              <h6
                className="MuiTypography-root MuiTypography-subtitle1"
                style={
                  Object {
                    "padding": "1rem 0",
                  }
                }
              >
                Implemented strategies
              </h6>
              <hr />
              <ul
                className="MuiList-root MuiList-padding"
              >
                <li
                  className="MuiListItem-root MuiListItem-gutters"
                  disabled={false}
                >
                  <div
                    className="MuiListItemAvatar-root"
                  >
                    <svg
                      aria-hidden={true}
                      className="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M20.5 11H19V7c0-1.1-.9-2-2-2h-4V3.5C13 2.12 11.88 1 10.5 1S8 2.12 8 3.5V5H4c-1.1 0-1.99.9-1.99 2v3.8H3.5c1.49 0 2.7 1.21 2.7 2.7s-1.21 2.7-2.7 2.7H2V20c0 1.1.9 2 2 2h3.8v-1.5c0-1.49 1.21-2.7 2.7-2.7 1.49 0 2.7 1.21 2.7 2.7V22H17c1.1 0 2-.9 2-2v-4h1.5c1.38 0 2.5-1.12 2.5-2.5S21.88 11 20.5 11z"
                      />
                    </svg>
                  </div>
                  <div
                    className="MuiListItemText-root MuiListItemText-multiline"
                  >
                    <span
                      className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                    >
                      <a
                        href="/strategies/view/StrategyA"
                        onClick={[Function]}
                      >
                        StrategyA
                      </a>
                    </span>
                    <p
                      className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                    >
                      A description
                    </p>
                  </div>
                </li>
                <li
                  className="MuiListItem-root MuiListItem-gutters"
                  disabled={false}
                >
                  <div
                    className="MuiListItemAvatar-root"
                  >
                    <svg
                      aria-hidden={true}
                      className="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M15.73 3H8.27L3 8.27v7.46L8.27 21h7.46L21 15.73V8.27L15.73 3zM12 17.3c-.72 0-1.3-.58-1.3-1.3 0-.72.58-1.3 1.3-1.3.72 0 1.3.58 1.3 1.3 0 .72-.58 1.3-1.3 1.3zm1-4.3h-2V7h2v6z"
                      />
                    </svg>
                  </div>
                  <div
                    className="MuiListItemText-root MuiListItemText-multiline"
                  >
                    <span
                      className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                    >
                      <a
                        href="/strategies/create"
                        onClick={[Function]}
                      >
                        StrategyB
                      </a>
                    </span>
                    <p
                      className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                    >
                      Missing, want to create?
                    </p>
                  </div>
                </li>
              </ul>
            </div>
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-md-12 MuiGrid-grid-xl-12"
            >
              <h6
                className="MuiTypography-root MuiTypography-subtitle1"
                style={
                  Object {
                    "padding": "1rem 0",
                  }
                }
              >
                1
                 Instances registered
              </h6>
              <hr />
              <ul
                className="MuiList-root MuiList-padding"
              >
                <li
                  className="MuiListItem-root MuiListItem-gutters"
                  disabled={false}
                >
                  <div
                    className="MuiListItemAvatar-root"
                  >
                    <svg
                      aria-hidden={true}
                      className="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M23 8c0 1.1-.9 2-2 2-.18 0-.35-.02-.51-.07l-3.56 3.55c.05.16.07.34.07.52 0 1.1-.9 2-2 2s-2-.9-2-2c0-.18.02-.36.07-.52l-2.55-2.55c-.16.05-.34.07-.52.07s-.36-.02-.52-.07l-4.55 4.56c.05.16.07.33.07.51 0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2c.18 0 .35.02.51.07l4.56-4.55C8.02 9.36 8 9.18 8 9c0-1.1.9-2 2-2s2 .9 2 2c0 .18-.02.36-.07.52l2.55 2.55c.16-.05.34-.07.52-.07s.36.02.52.07l3.55-3.56C19.02 8.35 19 8.18 19 8c0-1.1.9-2 2-2s2 .9 2 2z"
                      />
                    </svg>
                  </div>
                  <div
                    className="MuiListItemText-root MuiListItemText-multiline"
                  >
                    <span
                      className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                    >
                      instance-1 (4.0)
                    </span>
                    <p
                      className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                    >
                      <span>
                        123.123.123.123
                         last seen at
                         
                        <small>
                          23/02/2017, 15:56:49
                        </small>
                      </span>
                    </p>
                  </div>
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          aria-labelledby="wrapped-tab-1"
          hidden={true}
          id="wrapped-tabpanel-1"
          role="tabpanel"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly without permission 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  style={
    Object {
      "borderRadius": "10px",
      "boxShadow": "none",
    }
  }
>
  <div
    className="makeStyles-headerContainer-1"
  >
    <div
      className="makeStyles-headerTitleContainer-5"
    >
      <div
        className=""
        data-loading={true}
      >
        <h2
          className="MuiTypography-root makeStyles-headerTitle-6 MuiTypography-h2"
        >
          <span
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
              }
            }
          >
            <div
              className="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault"
              style={
                Object {
                  "marginRight": "8px",
                }
              }
            >
              <span
                aria-hidden={true}
                className="material-icons MuiIcon-root"
              >
                apps
              </span>
            </div>
            test-app
          </span>
        </h2>
      </div>
      <div
        className="makeStyles-headerActions-7"
      >
        <a
          aria-disabled={false}
          className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiButtonBase-root MuiIconButton-root MuiTypography-colorPrimary"
          href="http://example.org"
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z"
              />
            </svg>
          </span>
        </a>
      </div>
    </div>
  </div>
  <div
    className="makeStyles-bodyContainer-2"
  >
    <div>
      <p
        className="MuiTypography-root MuiTypography-body1"
      >
        app description
      </p>
      <p
        className="MuiTypography-root MuiTypography-body2"
      >
        Created: 
        <strong>
          Invalid Date
        </strong>
      </p>
    </div>
  </div>
</div>
`;
